# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [7]:
def double(num):
    return num * 2
print(double(4))

#syntax: Keywords Lambda, Variable(parameter): execution/transformation
print((lambda num: num * 2)(5))

# one-line execution is a must!

8
10


#### Saving to a Variable

In [9]:
lamb_test = lambda num: num * 2
print(lamb_test)
print(lamb_test(60))


<function <lambda> at 0x000001A907F38790>
120


#### Multiple Inputs

In [12]:
# AKA like destructuring in terms of how it looks.
# is actually 

print((lambda a, b, c : a  * b * c)(5, 14, 26))


mult = lambda a, b, c : a  * b * c
mult(5, 141, 26)

1820


18330

#### Passing a Lambda into a Function

In [16]:
# this function will take in a function!
def mult2(func, num):
    return func(num)

mult2(lambda x: x * x, 4)

mult2(lambda x: x ** x, 15)

437893890380859375

#### Returning a Lambda from a Function

In [23]:
# reg
def mult_test(num):
    return num * 5

# function in a function:
def returnFunk():
    def multiply(num):
        return num * 2
    return multiply

ret_test = returnFunk()
print(returnFunk())
print(ret_test)

def clo_test():
    names = []
    s = lambda name: names.append(name)
    t = lambda z : print(names)
    return s, t

add_name, print_info = clo_test()
print(add_name)
print(print_info)
add_name('Brendan')
print_info(';lakjsdf')

#last one I promise!
def return_of_lamb(x, y):
    return lambda a, b: x + y + a + b
ret_one =  return_of_lamb(78, 4)
print(ret_one)
print(type(ret_one))
print(ret_one)
print(ret_one(8, 64))
    

<function returnFunk.<locals>.multiply at 0x000001A909A2C9D0>
<function returnFunk.<locals>.multiply at 0x000001A909A2C700>
<function clo_test.<locals>.<lambda> at 0x000001A907F38700>
<function clo_test.<locals>.<lambda> at 0x000001A909A2D120>
['Brendan']
<function return_of_lamb.<locals>.<lambda> at 0x000001A907F38A60>
<class 'function'>
<function return_of_lamb.<locals>.<lambda> at 0x000001A907F38A60>
154


#### If Statements within Lambdas

In [24]:
#syntax:
# Lambda x: True if else False

cond_lamb = lambda num : num ** 2 if num < 10 else num -5

print(cond_lamb(5))
print(cond_lamb(10))
print(cond_lamb(15))

# Just in case, doubt this is useful and DO NOT pay too much attention to this
# and just a syntax example
# lambda x : x * 10 if x > 56 else (e **2 if x > 100 else x + 98)

25
5
10


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [32]:
f = lambda num: num ** 3

f(5)

125

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [42]:
# syntax --> map(function, iterable):
nums = [5, 87, 21, 4, 56, 89, 1]
nums2 = [4, 54, 21, 32, 3]

def square(num):
    return num * num

for n in range(len(nums)):
    nums[n] = square(nums[n])
print(nums)

#list comp
next_nums = [n * n for n in nums]
print(next_nums)

# time for the map func!

map(square, nums2) #This is our map object
for m in map(square, nums):
    print(m)

print(map(square, nums2))
list(map(square, nums2))

[25, 7569, 441, 16, 3136, 7921, 1]
[625, 57289761, 194481, 256, 9834496, 62742241, 1]
625
57289761
194481
256
9834496
62742241
1


[16, 2916, 441, 1024, 9]

#### Using Lambda's with Map

In [43]:
list(map(lambda x: x ** 2, nums2))

[16, 2916, 441, 1024, 9]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [47]:
list_of_nums = [3, 5, 6, 8]

list(map(lambda x: (x * 2) - 1, list_of_nums))

[5, 9, 11, 15]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [51]:
# syntax --> same as map.
# fliter(function, iterable)

nums3 = [87, 21, 2, 56, 1112, 24, 68]

def filter_odds(num):
    if num % 2 == 0:
        return True
    return False

print(list(filter(filter_odds, nums3)))

# string

name_list = ['Adele', 'Bill', 'Carter', 'Ava', 'Bob']

def a_name(name):
    if name[0].lower() == 'a':
        return True
    return False
        
list(filter(a_name, name_list))
    

[2, 56, 1112, 24, 68]


['Adele', 'Ava']


#### Using Lambda's with Filter()

In [53]:
list(filter(lambda name: name[0].lower() == 'a', name_list))

['Adele', 'Ava']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [60]:
from statistics import mean

nums3 = [87, 21, 2, 56, 1112, 24, 68]

def filter_mean(num):
    if num < mean(nums3):
        return True
    return False

list(filter(filter_mean, nums3))

[87, 21, 2, 56, 24, 68]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [64]:
from functools import reduce

#reduce(function, iterable)

# Reduce was intended to aggregate all items in a list like min(), max(), sum()
# Soooo, the function has to take 2 parameters and return 1

def add_up(num, num1):
    return num + num1
add_results = reduce(add_up, nums3)
print(add_results)

def subs(num, num1):
    return num1 - num
sub_results = reduce(subs, nums3)
print(sub_results)


1370
1168


#### Using Lambda's with Reduce()

In [65]:
lamb_res = reduce(lambda x, y: x + y, nums3)
print(lamb_res)

1370


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [ ]:
#skip

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [68]:

# def addNums(num):
#    return  num + addNums(num + 1)
# be careful to increment or decrement - a call here would be an infitine loop!

def addNums(num):
    if num >= 10:
        #print('addNums(10) = 10')
        return  num
    else:
        #print(f"addNums({num}) = {num} = addNums({num + 1})")
        return num + addNums(num + 1)
addNums(1)

55

#### Writing a Factorial Function

In [71]:
#while loop version first
def  factor(num):
    count = 1
    while num >=1:
        print(count, num)
        count = num * count
        num -= 1
    return count
print(factor(5))

def factor2(num):
    if num <= 1:
        print('reached base case')
        return 1
    else:
        print(num)
        return num * factor2(num - 1)
    
print(factor2(5))

1 5
5 4
20 3
60 2
120 1
120
5
4
3
2
reached base case
120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [9]:
def subNums(num):
    if num <= 1:
        print('subNums(1) = 1')
        return  num
    else:
        print(f"subNums({num}) = {num} - subNums({num - 1})")
        return num - subNums(num - 1)
subNums(5)

subNums(5) = 5 - subNums(4)
subNums(4) = 4 - subNums(3)
subNums(3) = 3 - subNums(2)
subNums(2) = 2 - subNums(1)
subNums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [14]:
# yield is kinda like return, but keep returning. . .

range(1,11)
print(range(1,11))
print(type(range(1,11)))
print(list(range(1,11)))
#indexing example
print(range(1,11)[2])

# the only data stored for a generator is the formula to run when it's called on

range(1, 11)
<class 'range'>
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
3


In [ ]:
def my_range(start, stop, step):
    while start < stop:
        yield start
        start += step
        
print(list(my_range(0, 10, 2)))

for x in my_range(0, 10, 2):
    print(x)

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [24]:
def new_squares(nums):
    while nums >= 0:
        yield nums ** 2
        nums -= 1
for s in new_squares(10):
    print(s)

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [144]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def filter_empty(place):
    if place.strip() == '':
        return False
    return True

print(list(filter(filter_empty, places)))

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [142]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key = lambda x: x.split()[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [140]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

list(map(lambda x: (x[0], (9/5) * x[1] + 32), places))

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [214]:
def sequence(num):
    if num < 2:
        return 1
    else:
        return sequence(num - 2) + sequence(num - 1)

def fib(iteration):
    for x in range(iteration + 1):
        print(f"Iteration {x}: {sequence(x)}")
        
fib(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
